connectors/services/job_executor.py

```diff
@@ -0,0 +1,228 @@
+
+        data_provider = source_klass(configuration)
+
+        if not await data_provider.changed():
+            logger.debug(f"No change in {service_type} data provider, skipping...")
+            return
+
+        logger.debug(f"Syncing '{service_type}'")
+        try:
+            logger.debug(f"Pinging the {source_klass} backend")
+            await data_provider.ping()
+            await asyncio.sleep(0)
+
+            mappings, settings = defaults_for(
+                is_connectors_index=True, language_code=language_code
+            )
+            logger.debug("Preparing the index")
+            await self.elastic_server.prepare_index(
+                index_name, mappings=mappings, settings=settings
+            )
+            await asyncio.sleep(0)
+
+            result = await self.elastic_server.async_bulk(
+                index_name,
+                self.prepare_docs(pipeline, data_provider),
+                pipeline,
+                options=self.bulk_options,
+            )
+            for i in range(50):
+                await asyncio.sleep(0.1)
+            await self._sync_done(result, start_time)
+
+        except Exception as e:
+            await self._sync_done({}, start_time, exception=e)
+            raise
+        finally:
+            self._start_time = None
+
+    async def prepare_docs(self, pipeline, data_provider):
+        logger.debug(f"Using pipeline {pipeline}")
+
+        async for doc, lazy_download in data_provider.get_docs():
+            # adapt doc for pipeline settings
+            doc["_extract_binary_content"] = pipeline.extract_binary_content
+            doc["_reduce_whitespace"] = pipeline.reduce_whitespace
+            doc["_run_ml_inference"] = pipeline.run_ml_inference
+            yield doc, lazy_download
+
+    async def _sync_done(self, result, start_time, exception=None):
+        doc_updated = result.get("doc_updated", 0)
+        doc_created = result.get("doc_created", 0)
+        doc_deleted = result.get("doc_deleted", 0)
+        exception = result.get("fetch_error", exception)
+
+        indexed_count = doc_updated + doc_created
+
+        await self.sync_job.done(indexed_count, doc_deleted, exception)
+
+        # Loading connector as late as possible so that there's less likely
+        # a version conflict.
+        connector = await self.connector_index.fetch_by_id(self.sync_job.connector_id)
+        await connector.job_done(self.sync_job)
+
+        logger.info(
+            f"Sync done: {indexed_count} indexed, {doc_deleted} "
+            f" deleted. ({int(time.time() - start_time)} seconds)"
+        )
+
+
+class JobExecutorService(BaseService):
+    def __init__(self, config):
+        super().__init__(config)
+        self.errors = [0, time.time()]
+        self.service_config = self.config["service"]
+        self.idling = self.service_config["idling"]
+        self.connector_index = None
+        self.sync_job_index = None
+
+    def raise_if_spurious(self, exception):
```

Comment:
this can go into the `BaseService` class along with `self.errors` -- we are likely to use it in all services.

Reply:
I have done it in this PR: https://github.com/elastic/connectors-python/pull/337